internal/api/kubernetes/client.go

```diff
@@ -510,93 +522,174 @@ func gvrAndKeyFromObj(
+			}
+			for _, sa := range serviceAccounts {
+				_, err := getUserClient(
+					ctx,
+					internalClient.Scheme(),
+					ra,
+					withServiceAccount(sa.Namespace, sa.Name),
+				)
+				if err == nil {
+					// Note: At present, we never return the user-specific client. We always
+					// return the internal client so that most operations are performed by a
+					// single client with a built-in cache.
+					return internalClient, nil
+				}
+				if !apierrors.IsForbidden(err) {
+					return nil, errors.Wrap(err, "get service account specific client")
+				}
+			}
+			return nil, newForbiddenError(ra)
+		}
+
+		// If we get to here, we're dealing with a user who "authenticated" by just
+		// passing their bearer token for the Kubernetes API server.
+
+		// Calling getUserClient will build a user-specific client and conduct a
+		// SelfSubjectAccessReview. If the review succeeds, we'll have obtained a
+		// client that can perform the desired operation.
+		_, err := getUserClient(
+			ctx,
+			internalClient.Scheme(),
+			ra,
+			withBearerToken(userInfo.BearerToken),
+		)
+		if err != nil {
+			return nil, errors.Wrap(err, "get user-specific client")
+		}
+		// Note: At present, we never return the user-specific client. We always
+		// return the internal client so that most operations are performed by a
+		// single client with a built-in cache.
+		return internalClient, nil
+	}
+}
+
+type userClientOption func(*userClientOptions)
+
+func withBearerToken(bearerToken string) userClientOption {
+	return func(opts *userClientOptions) {
+		opts.bearerToken = bearerToken
 	}
-	if client == nil {
-		return nil, errors.New("not allowed")
+}
+
+func withServiceAccount(namespace, name string) userClientOption {
+	return func(opts *userClientOptions) {
+		opts.subject = &userClientSubject{
+			username: fmt.Sprintf("system:serviceaccount:%s:%s", namespace, name),
+		}
 	}
-	// Note: At present, we never return the user-specific client. We always
-	// return the internal client so that most operations are performed by a
-	// single client with a built-in cache.
-	return internalClient, nil
+}
+
+type userClientSubject struct {
+	username string
+}
+
+type userClientOptions struct {
+	bearerToken string
+	subject     *userClientSubject
 }
 
 // getUserClient creates a user-specific client using the provided scheme and
-// bearer token. It then uses that client to conduct a SelfSubjectAccessReview
-// to determine whether the token provided permits the desired operation. If so,
-// the user-specific client is returned. Otherwise, nil is returned.
+// options. It then uses that client to submit a SelfSubjectAccessReview
+// to determine whether the token provided permits the desired operation.
+// If so, the user-specific client is returned.
+//
+//nolint:unparam
 func getUserClient(
 	ctx context.Context,
 	scheme *runtime.Scheme,
-	bearerToken string,
-	resourceAttrs authv1.ResourceAttributes,
+	ra authv1.ResourceAttributes,
+	opts ...userClientOption,
```

Comment:
I know I wrote this originally...

I must have thought at one time that we might potentially _use_ the user-specific client, but we don't and it seems clear that we never will.

We should probably change the name of the function and return types to reflect that this is just carrying out some access review. There's no need to return the client.